### PR TITLE
kernelci.legacy: Install vmlinux to aid debugging

### PR DESCRIPTION
**Pull request by broonie**
Some debugging, such as the use of scripts/decode_stacktrace.sh and other things involving looking at the disassembly of the built kernel, need the vmlinux to get debug information which doesn't make it into the image files deployed in onto targets for most architectures. Support such uses by including the vmlinux in our shipped artefacts.

Since this is frequently a large file but compresses well use xz to compress as we install. For example, a build I have here is 176M but with xz we can compress it down to 52M. gzip is much less effective at 80M although it is faster, the tradeoff seems worth it given that the binaries are likely to be used infrequently.

Signed-off-by: Mark Brown <broonie@kernel.org>

### CODE REVIEW

**Comment by broonie** [comment]
File present on staging: https://storage.staging.kernelci.org/kernelci/staging-mainline/staging-mainline-20230115.0/arm64/defconfig/gcc-10/kernel/ and listed in the YAML: https://storage.staging.kernelci.org/kernelci/staging-mainline/staging-mainline-20230115.0/arm64/defconfig/gcc-10/artifacts.json

**Comment by gctucker** [comment]
> kernelci.legacy: Install vmlinux to aid debugging

This change is actually in `kernelci.build`.

**Comment by gctucker** [comment]
> Since this is frequently a large file but compresses well use xz to compress as we install. For example, a build I have here is 176M but with xz we can compress it down to 52M. gzip is much less effective at 80M although it is faster, the tradeoff seems worth it given that the binaries are likely to be used infrequently.

Yes I think this is the actual reason why it's not sent to storage at the moment.  If we do that, we need to extend our storage disk by a significant amount (to be determined) or reduce the garbage collection period.

**Comment by gctucker** [comment]
Maybe a reasonable compromise would be to only store vmlinux on the defconfig builds rather than all the builds?

**Comment by broonie** [comment]
That might work - I'd suggest also including the kselftest builds in there given what it's testing, there are some kselftests that explicitly create oopses. We could also do it per tree, for example focusing on mainline and next. We could also look into something like having a faster expiration policy for vmlinux files - probably if someone hasn't looked at a vmlinux for say a week it's probably safer to delete it for example - or having cold storage in the Azure equivalent of a S3 bucket that we shuffle things off to.

**Comment by gctucker** [comment]
A dedicated service with the new pipeline could be using any separate arbitrary storage for that.

OK so I guess we could add an optional parameter to the build configs e.g.:
```yaml
params:
    upload_vmlinux: true
```
for now?  Then the code in `build.py` would use that to implement its `is_enabled()` method.  It's not perfect design but it would be not too intrusive and a fair compromise I believe.

**Comment by broonie** [comment]
Yes, sounds OK for now - we can remove it later if we get less constrained on storage one way or another.

File: kernelci/build.py [hunk 1]
```diff
@@ -1286,6 +1286,21 @@ def _install_system_map(self, kbmeta, verbose):
...
 
+    def _install_vmlinux(self, kbmeta, verbose):
+        file_name = 'vmlinux'
+        vmlinux = os.path.join(self._output_path, file_name)
+        file_name_xz = 'vmlinux.xz'
+        vmlinux_xz = os.path.join(self._output_path, file_name_xz)
+        if os.path.exists(vmlinux):
+            if verbose:
+                print("Creating {}".format(file_name_xz))
+            shell_cmd("xz -T0 --compress < {src} > {dest}".format(
```
Comment (nuclearcat) [inline kernelci/build.py:1297]: Probably you can use same approach as we have already in code? (i know it ingest whole file in RAM, but i think it is fine with size we have):
https://github.com/kernelci/kernelci-core/blob/4c6c5065c4e9a0b57bbedc148db6edc517876764/kci_build#L373

Comment (broonie) [reply kernelci/build.py:1297]: I can't tell what you're suggesting here, sorry.

**Comment by nuclearcat** [comment]
This patch causes significant storage usage increase:
root@kernelci-staging:/data/storage.staging.kernelci.org/next# du --max-depth=1 -h
10G	./master
10G	.
root@kernelci-staging:/data/storage.staging.kernelci.org/next# find . -name "vmlinux.xz" -print0 | xargs -0 du -ch | tail -n 1
3.6G	total

So from total 10G it takes 3.6G.

**Comment by nuclearcat** [comment]
Extract from KCI weekly:
We need to do selective builds and build vmlinux only one arm, one for x86 (chromebook+kselftest+etc+some config options for debugging trace), and one for arm64.

**Comment by gctucker** [comment]
@broonie Could you please update this PR to only store `vmlinux` on the small subset of builds as per the comment above?

**Comment by broonie** [comment]
If we are switching to object storage we may want to revisit saving only sometimes, it should vastly reduce costs.

**Comment by nuclearcat** [comment]
This patch likely causing breakage on staging:
```
Traceback (most recent call last):
  File "/usr/local/bin/kci_build", line 4, in <module>
    __import__('pkg_resources').run_script('kernelci==1.1', 'kci_build')
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 651, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/usr/lib/python3/dist-packages/pkg_resources/__init__.py", line 1448, in run_script
    exec(code, namespace, namespace)
  File "/usr/local/lib/python3.9/dist-packages/kernelci-1.1-py3.9.egg/EGG-INFO/scripts/kci_build", line 480, in <module>
    status = opts.command(configs, opts)
  File "/usr/local/lib/python3.9/dist-packages/kernelci-1.1-py3.9.egg/EGG-INFO/scripts/kci_build", line 304, in __call__
    install_res = self._install_step(step, configs, args)
  File "/usr/local/lib/python3.9/dist-packages/kernelci-1.1-py3.9.egg/EGG-INFO/scripts/kci_build", line 317, in _install_step
    return step.install(args.verbose, configs)
TypeError: install() takes from 1 to 2 positional arguments but 3 were given
Build result: 1
```
Still investigating, will edit this post after run completed.

**Comment by nuclearcat** [comment]
Closing, as we are deprecating the legacy system. If this is still an issue or relevant, please re-open it under the Maestro system. Thank you for your contribution, and sorry for the inconvenience.

**Pull request closed without merging**